plugins/outputs/cloud_pubsub/pubsub.go

```diff
@@ -241,8 +268,20 @@ func (ps *PubSub) toMessages(metrics []telegraf.Metric) ([]*pubsub.Message, erro
 			b = []byte(encoded)
 		}
 
+		b, err = ps.encoder.Encode(b)
+		if err != nil {
+			ps.Log.Debugf("could not encode metric: %v", err)
+			continue
+		}
+		var data []byte
+		if ps.ContentEncoding == "gzip" {
+			data = make([]byte, len(b))
+			copy(data, b)
+		} else {
+			data = b
+		}
```

Comment:
See above, I think you don't need to copy here.

Reply:
IMHO the copy is needed because GzipDecoder [Decode ](https://github.com/influxdata/telegraf/blob/master/internal/content_coding.go#L169) is returning [buffer.Bytes() ](https://pkg.go.dev/bytes#Buffer.Bytes)which is a [slice](https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/bytes/buffer.go;l=54) and we are appending [here](https://github.com/influxdata/telegraf/blob/master/plugins/outputs/cloud_pubsub/pubsub.go#L244-L245) and this causes erroneous result.  I have added [couple of tests](https://github.com/influxdata/telegraf/pull/9159/files#diff-c333930dc7e05efd47fb16f1fc53057bc5f9407e36851f5692accd06729e5e98R168-R223) to validate that. If we remove the copy those tests starts failing

Reply:
Well but you are `Encode` here, don't you? Anyway, if you are sure you need the copy, please feel free to close this conversation as resolved.

Reply:
Oops I do mean to say GzipEncoder [Encode ](https://github.com/influxdata/telegraf/blob/master/internal/content_coding.go#L125) is returning [buffer.Bytes() ](https://pkg.go.dev/bytes#Buffer.Bytes)which is a [slice](https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/bytes/buffer.go;l=54) and we are appending [here](https://github.com/influxdata/telegraf/blob/master/plugins/outputs/cloud_pubsub/pubsub.go#L244-L245) and this causes erroneous result.